Issue 2452 Add X-Debug-Token Passive Scanner

kingthorin opened the pull request:
XDebugTokenScanner - The new scanner.
Messages.properties - Add necessary i18n keys and values.
pscanalpha.html - Add new help entry.
ZapAddOn.xml - Add new change entry, and include the new scanner in the
pscanrules section.
XDebugTokenScannerUnitTest - The unit tests for the new scanner.

Fixes zaproxy/zaproxy#2452

psiinon commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/help/contents/pscanalpha.html:
```diff
@@ -162,5 +162,9 @@
 
+<H2>X-Debug-Token Information Leak</H2>
+This checks response headers for the presence of X-Debug-Token and X-Debug-Token-Link details.
+Which suggest the use/exposure of Symfony's Profiler, which can leak a significant amount of information of interest to Security Testers and malicious individuals.
```
Nit: "Which suggest.. , which can leak.." - the 2 'which's dont read quite right to me. 
Everything else looks good though :)

kingthorin replied:
Good point. For the first one "Which indicates..." ? For the second one I'll remove "can", there's no reason to be non-commital, it really does happen.

kingthorin replied:
Tweaked help and messages, to make things less wishy-washy

kingthorin commented:
@thc202

kingthorin commented:
Added unit test.

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XDebugTokenScanner.java:
```diff
@@ -0,0 +1,170 @@
+		Vector<String> xdtlHeaders = msg.getResponseHeader().getHeaders(X_DEBUG_TOKEN_LINK_HEADER);
+		if (xdtlHeaders != null) { // Header Found
+			for (String xdtl : xdtlHeaders) {
+				parent.raiseAlert(id, getAlert(msg, X_DEBUG_TOKEN_LINK_HEADER));
```
The value of the header is not being used (the link is being constructed with the header name instead of the value).

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XDebugTokenScanner.java:
```diff
@@ -0,0 +1,170 @@
+		Vector<String> xdtHeaders = msg.getResponseHeader().getHeaders(X_DEBUG_TOKEN_HEADER);
+		if (xdtHeaders != null) { // Header Found
+			for (String xdt : xdtHeaders) {
+				parent.raiseAlert(id, getAlert(msg, X_DEBUG_TOKEN_HEADER));
```
- Should break/return here as well? (in case of multiple headers)
 - Same comment regarding the header value.

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XDebugTokenScanner.java:
```diff
@@ -0,0 +1,170 @@
+		if (assembledURI.isAbsoluteURI()) { // Full URL: http://example.com/_profiler/3d02cb
+			return assembledURI.toString();
+		} else if (headerValue.toLowerCase().contains(PROFILER_PORTION)) { // Full Path: /_profiler/3d02cb 
+			URI currentURI = msg.getRequestHeader().getURI();
```
It could be extracted a method that accepts the current URI and a path, it would be used by both ifs (full path/token only).

Worth adding tests for these cases too.

thc202 replied:
Regarding adding the tests, it's just missing the full path (token is already being tested).

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties:
```diff
@@ -248,3 +248,9 @@ pscanalpha.xchromeloggerdata.refs=https://craig.is/writing/chrome-logger
+pscanalpha.xdebugtoken.name=X-Debug-Token Information Leak
+pscanalpha.xdebugtoken.desc=The response contained an X-Debug-Token or X-Debug-Token-Link header. This indicates that Symfony's Profiler may be in use and exposing sensitive data.
+pscanalpha.xdebugtoken.otherinfo=By accessing: {0}, you may gain access to the profiler and further leaked information.\n(This value was assembled based on an educated guess, there is no guarantee that it is correct or will work as expected). 
+pscanalpha.xdebugtoken.refs=http://symfony.com/doc/current/cookbook/profiler/profiling_data.html\nhttp://symfony.com/blog/new-in-symfony-2-4-quicker-access-to-the-profiler-when-working-on-an-api
```
It could use HTTPS.

thc202 commented on test/org/zaproxy/zap/extension/pscanrulesAlpha/XDebugTokenScannerUnitTest.java:
```diff
@@ -0,0 +1,103 @@
+
+		// Then
+		assertThat(alertsRaised.size(), is(1));
+		assertThat(alertsRaised.get(0).getEvidence(), is(X_DEBUG_TOKEN_HEADER));
```
Worth asserting that the other info contains the expected URL (per previous comments about the URL not being the expected one).

kingthorin commented:
I simplified the other info part, I'm sure users can figure out assembling a URL on their own with an example. I also changed one of the tests to use the partial URL version of the header.

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/XDebugTokenScanner.java:
```diff
@@ -0,0 +1,162 @@
+	 * @param header the name of the header field being looked for
+	 * @return boolean status of existence
+	 */
+	private boolean requestHasHeader(HttpMessage msg, String header) {
```
It should be `response...`.

kingthorin replied:
Done

thc202 commented on src/org/zaproxy/zap/extension/pscanrulesAlpha/resources/Messages.properties:
```diff
@@ -248,3 +248,9 @@ pscanalpha.xchromeloggerdata.refs=https://craig.is/writing/chrome-logger
+pscanalpha.xdebugtoken.name=X-Debug-Token Information Leak
+pscanalpha.xdebugtoken.desc=The response contained an X-Debug-Token or X-Debug-Token-Link header. This indicates that Symfony's Profiler may be in use and exposing sensitive data.
+pscanalpha.xdebugtoken.otherinfo=By accessing a URL in the form http://target_host/_profiler/token_value (i.e.: http://example.com/_profiler_/123ab4), you may gain access to the profiler and further leaked information.
+pscanalpha.xdebugtoken.refs=http://symfony.com/doc/current/cookbook/profiler/profiling_data.html\nhttp://symfony.com/blog/new-in-symfony-2-4-quicker-access-to-the-profiler-when-working-on-an-api
```
This was not changed to HTTPS?

kingthorin replied:
Done, thanks

thc202 commented:
Thanks!

kingthorin commented:
@psiinon still good with this?

psiinon commented:
Yeah :)
